In [2]:
import pandas as pd
import numpy as np
import math
from ast import literal_eval
import random

## 10 Points
## 10 Puntos

In [2]:
def f(x1,x2):
    return (1.5-x1+x1*x2)**2+(2.25-x1+x1*x2**2)**2+(2.625-x1+x1*x2**3)**2

In [3]:
def RealF(x1):
    
    value = f(x1[0],x1[1])
    return 1/(value+1)
    

In [4]:
def num_bits(bounds):
    return math.ceil(math.log((bounds[1]-(bounds[0]))*10**3,2))


In [5]:
num_bits([-4.5,4.5])

14

In [6]:
def initialPopu(bits,popu,variables):
    Pupulation =[]
    for i in range(popu):
        individual = ""
        for i in range(bits*variables):
            if np.random.uniform(0,1,1)[0] < 0.5:
                #individual.append(0)
                
                individual = individual + "0"
            else:
                #individual.append(1)
                individual = individual + "1"
        Pupulation.append(individual)
    return Pupulation
            

In [7]:
initial = initialPopu(14,100,2)
initial

['0000001110001010001000010011',
 '0110100101000111001001100111',
 '0100111110000000111010000101',
 '0001000000110101110100011011',
 '1100001011110010111010011101',
 '0011110100010110101000101010',
 '0101100101110111100110100011',
 '1101001110111001100101001001',
 '1001100100011100011101100000',
 '1101101101110001100011010110',
 '1010101011010000011011010011',
 '0000101111010000111110101011',
 '0010100000001000110001001101',
 '1111111010110000111010011101',
 '1110011101110010011010110101',
 '0000100011100010110011011011',
 '0111100011110000011000100011',
 '1000101011010001011111010000',
 '0100000011101010110110101000',
 '0101101000110111001000111011',
 '1100110101001101000111011000',
 '0001001001110001101101010010',
 '0000101011010111110110111000',
 '1101111011100100111011110110',
 '0100101001000000101000111101',
 '0010101110101101001111011010',
 '1101111100010000110110000000',
 '1111100001110111001110100010',
 '0010010011100101101101011001',
 '0101011010100101001000001101',
 '00001011

In [8]:
def getValues(chromosomes,num_bits, variables, bounds):
    c=0
    curr = ""
    valuesR = []
    valuesN = []
    for chromosome in range(len(chromosomes)):
        for i in chromosomes[chromosome]:
            curr = curr+i
            c+=1
            if c == num_bits:
                x = int(curr,base=2)
                valuesN.append(x)
                newx = bounds[0] + ((9*x)/((2**num_bits)-1))
                valuesR.append(newx)
                curr = ""
                c = 0
        chromosomes[chromosome] = [chromosomes[chromosome],valuesN,valuesR,f(valuesR[0],valuesR[1]),RealF(valuesR)]
        valuesR = []
        valuesN = []
    
    return chromosomes

In [9]:
def getOneValues(chromosomes,num_bits, variables, bounds):
    c=0
    curr = ""
    valuesR = []
    valuesN = []

    for i in chromosomes[0]:
        curr = curr+i
        c+=1
        if c == num_bits:
            x = int(curr,base=2)
            valuesN.append(x)
            newx = bounds[0] + ((9*x)/((2**num_bits)-1))
            valuesR.append(newx)
            curr = ""
            c = 0
    print(valuesR)
    new_chromosome = [chromosomes[0],valuesN,valuesR,f(valuesR[0],valuesR[1]),RealF(valuesR)]
    valuesR = []
    valuesN = []
    
    return new_chromosome

In [10]:
getOneValues(['1100000000010000101101101110'], 14, 2, [-4.5, 4.5])

[2.252609412195569, -2.8926020875297564]


['1100000000010000101101101110',
 [12292, 2926],
 [2.252609412195569, -2.8926020875297564],
 3339.887797584166,
 0.0002993216356212595]

In [11]:
initial2 = getValues(initial, 14, 2, [-4.5, 4.5])
initial2


[['0000001110001010001000010011',
  [226, 8723],
  [-4.3758469144845265, 0.2919794909357263],
  107.73943385588002,
  0.009196295810454282],
 ['0110100101000111001001100111',
  [6737, 12903],
  [-0.7990294817798937, 2.5882622230360743],
  114.1531576166626,
  0.008684086660731746],
 ['0100111110000000111010000101',
  [5088, 3717],
  [-1.704907526094122, -2.4580662882256],
  974.1505140128991,
  0.001025482718441937],
 ['0001000000110101110100011011',
  [1037, 7451],
  [-3.9303241164621863, -0.40679362754074333],
  126.50018442568052,
  0.00784312591000915],
 ['1100001011110010111010011101',
  [12476, 11933],
  [2.353689800402857, 2.055392785204175],
  541.5608722741694,
  0.0018431111624552892],
 ['0011110100010110101000101010',
  [3909, 10794],
  [-2.3525911005310385, 1.4296832082036257],
  3.8814746473152892,
  0.20485612898757538],
 ['0101100101110111100110100011',
  [5725, 14755],
  [-1.3549716169199781, 3.6056583043398653],
  3744.9750998516033,
  0.0002669531893150104],
 ['110100

In [12]:
arr = getValues(['0111010000100011010101011001'], 14, 2, [-4.5, 4.5])
arr

[['0111010000100011010101011001',
  [7432, 13657],
  [-0.41723127632301793, 3.0024720747115916],
  69.71559241216701,
  0.014141152833331061]]

In [13]:
def getMean(chromosomes,mu):
    mean = 0
    for i in chromosomes:
        mean += i[-1]
    return mean/mu
    

In [14]:
getMean(initial2,100)

0.02209000171714478

In [15]:
initial2[0][-1]

0.009196295810454282

In [16]:
def parentsSelection(chromosomes,mu):
    r = random.random()
    mean = getMean(chromosomes,mu)
    cont = 0
    parents = []
    for i in range(mu):
        cont += (chromosomes[i][-1] / mean )
        while cont > r:
            parents.append(chromosomes[i])
            r+=1
    return parents

In [17]:
arr = parentsSelection(initial2,100)
arr

[['0000001110001010001000010011',
  [226, 8723],
  [-4.3758469144845265, 0.2919794909357263],
  107.73943385588002,
  0.009196295810454282],
 ['0001000000110101110100011011',
  [1037, 7451],
  [-3.9303241164621863, -0.40679362754074333],
  126.50018442568052,
  0.00784312591000915],
 ['0011110100010110101000101010',
  [3909, 10794],
  [-2.3525911005310385, 1.4296832082036257],
  3.8814746473152892,
  0.20485612898757538],
 ['0011110100010110101000101010',
  [3909, 10794],
  [-2.3525911005310385, 1.4296832082036257],
  3.8814746473152892,
  0.20485612898757538],
 ['0011110100010110101000101010',
  [3909, 10794],
  [-2.3525911005310385, 1.4296832082036257],
  3.8814746473152892,
  0.20485612898757538],
 ['0011110100010110101000101010',
  [3909, 10794],
  [-2.3525911005310385, 1.4296832082036257],
  3.8814746473152892,
  0.20485612898757538],
 ['0011110100010110101000101010',
  [3909, 10794],
  [-2.3525911005310385, 1.4296832082036257],
  3.8814746473152892,
  0.20485612898757538],
 ['001

In [18]:
def combination(chro1,chro2, prob):
    r = np.random.uniform(0,1,1)[0]
    if prob > r :
        
    
        ind1 = random.randint(1,len(chro1[0]))
        ind2 = random.randint(1,len(chro1[0]))

        if ind1 == ind2:
            while True:
                ind2 = random.randint(1,len(chro1[0]))
                if ind1 != ind2:
                    break
        if ind2 < ind1: 
            ind2 = ind1
            ind1 = ind2

        new1 = "" 
        new2 = ""
        new1 = new1 + chro1[0][:ind1]
        new2 = new2 + chro2[0][:ind1]

        new1 = new1 + chro2[0][ind1:ind2]
        new2 = new2 + chro1[0][ind1:ind2]

        new1 = new1 + chro1[0][ind2:]
        new2 = new2 + chro2[0][ind2:]
        new1 = getValues([new1], 14, 2, [-4.5, 4.5])
        new2 = getValues([new2], 14, 2, [-4.5, 4.5])

        return new1,new2
    
    else: 
        return chro1,chro2,
        
    

In [19]:
combination(initial2[0],initial2[1],0.9)

([['0010100101000111001000010011',
   [2641, 12819],
   [-3.049166819263871, 2.5421168284197035],
   2190.7134256735526,
   0.00045626402990741466]],
 [['0100001110001010001001100111',
   [4322, 8807],
   [-2.125709577000549, 0.3381248855520971],
   47.324654243309496,
   0.020693371026828382]])

In [20]:
def mutation(arr,pm):
    chrom = arr[0]
    newchrom = ""
    for i in chrom:
        if np.random.uniform(0,1,1)[0] < pm:
            if i == '0':
                newchrom += '1'
            else:
                newchrom += '0'
        else:
            newchrom += str(i)
    return getValues([newchrom], 14, 2, [-4.5, 4.5])

In [21]:
mutation(['1101100001111001111000101010',
   [13854, 7722],
          
   [3.1106940120856983, -0.25791979490935724],
   0.1326157314831691],0.0001)

[['1101100001111001111000101010',
  [13854, 7722],
  [3.1106940120856983, -0.25791979490935724],
  6.540583525167335,
  0.1326157314831691]]

In [22]:
def beestworst(arr):
    best = arr[0][-1]
    best_complete_array = arr[0]
    worst =arr[0][-1]
    worst_complete_array = arr[0]
    for i in arr:
        if i[-1] > best:
            best = i[-1]
            best_complete_array = i
            
        if i[-1] < worst:
            worst = i[-1]
            worst_complete_array = i
    return best_complete_array,worst_complete_array

In [23]:
def survivorseries(parents, son):
    bestB, worstB =  beestworst(parents)
    bestW, worstW =  beestworst(son)
    if bestB[-1] > worstW[-1]:
        son.remove(worstW)
        son.append(bestB)
    return son

In [24]:
def AG(G,mu,pc,pm,bounds,variables):
    bits = num_bits(bounds)
    initial = initialPopu(bits,mu,variables)
    initial = getValues(initial,bits, variables, bounds)
    t = 0
    while t<=G:
        parents = parentsSelection(initial,mu)
        sons = []
        
        for i in range(0,mu,2):
            if random.random() < pc:
                h1,h2=  combination(parents[i],parents[i+1],0.9)
            else:
                h1 = parents[i].copy()
                h2 = parents[i+1].copy()
            if len(h1[0][0]) == 1:
                h1 = [h1]
                
            if len(h2[0][0]) == 1:
                h2 = [h2]
            h1 = mutation(h1[0],pm)
            h2 = mutation(h2[0],pm)
            sons.append(h1[0])
            sons.append(h2[0])
        
        initial = survivorseries(parents, sons)
        t +=1
        
    best,_ = beestworst(initial)  
    return best
    

In [26]:
AG(100,100,0.9,0.0001,[-4.5, 4.5],2)

['1101010100011110001101111010',
 [13639, 9082],
 [2.992583775865226, 0.48919611792711937],
 0.0018145106539843795,
 0.998188775831566]

## 10 Points
## 10 Puntos

In [25]:
def initialPopuT(popu,variables, bounds):
    Pupulation =[]
    for i in range(popu):
        
        b1 = np.random.uniform(bounds[0],bounds[-1],variables)
        
        Pupulation.append(b1)
    return Pupulation
            
            

In [26]:
initial = initialPopuT(100,2,[-4.5,4.5])
initial

[array([0.83071645, 0.06269388]),
 array([ 4.24385979, -0.9577582 ]),
 array([2.53761963, 2.59899801]),
 array([ 3.16305887, -1.64684452]),
 array([ 2.45049535, -4.00867168]),
 array([3.07897538, 2.12862148]),
 array([-1.72388929,  4.16682782]),
 array([-3.30335957,  3.035622  ]),
 array([ 2.80762926, -4.40797031]),
 array([-2.65623267,  0.03454488]),
 array([3.752152  , 0.87874363]),
 array([-1.15735031,  1.2734984 ]),
 array([-2.15340045,  4.15676595]),
 array([ 1.15829228, -1.74940018]),
 array([2.9732235 , 3.36605015]),
 array([ 3.35375383, -3.64998892]),
 array([2.16526153, 0.73213373]),
 array([3.86336517, 2.61440237]),
 array([-0.80553027,  0.52155068]),
 array([2.03379395, 3.32264081]),
 array([-3.08780638, -3.04910597]),
 array([-2.92642894,  2.90939026]),
 array([0.26564829, 2.58568815]),
 array([-0.91481968, -1.34220033]),
 array([-0.47011774, -3.39308221]),
 array([ 0.31736799, -0.0649766 ]),
 array([-2.23213686, -0.08623271]),
 array([-0.97539432, -1.42272397]),
 array([-0

In [27]:
def getValuesT(chromosomes, variables, bounds):
    c=0
    valuesR = []
    valuesN = []
    for i in range(len(chromosomes)):

#        newx = bounds[0] + ((9*chromosomes[0])/((2**num_bits)-1))
#        valuesR.append(newx)
            
#            curr = ""
#            c = 0
     chromosomes[i]= [chromosomes[i],f(chromosomes[i][0],chromosomes[i][1]),RealF(chromosomes[i])]
#    valuesR = []
#    valuesN = []
    
    return chromosomes

In [28]:

initial2 = getValuesT(initial, 2, [-4.5,4.5])
initial2

[[array([0.83071645, 0.06269388]), 5.764199474317527, 0.14783715409293055],
 [array([ 4.24385979, -0.9577582 ]), 78.55510025844417, 0.012569904339902552],
 [array([2.53761963, 2.59899801]), 2307.392048910704, 0.00043320197731225296],
 [array([ 3.16305887, -1.64684452]), 321.0638131642907, 0.0031049747258934724],
 [array([ 2.45049535, -4.00867168]), 26513.75376768989, 3.771485146577416e-05],
 [array([3.07897538, 2.12862148]), 1052.0465901573896, 0.0009496255999941454],
 [array([-1.72388929,  4.16682782]),
  15177.953077668522,
  6.588069644086411e-05],
 [array([-3.30335957,  3.035622  ]),
  8124.998562319974,
  0.00012306179878457914],
 [array([ 2.80762926, -4.40797031]),
  61015.031664256974,
  1.638913532598347e-05],
 [array([-2.65623267,  0.03454488]), 68.45023151559613, 0.014398800092918832],
 [array([3.752152  , 0.87874363]), 5.052004191276636, 0.1652345187469303],
 [array([-1.15735031,  1.2734984 ]), 5.680287107391992, 0.14969416492495688],
 [array([-2.15340045,  4.15676595]),
  2

In [29]:
def getMeanT(chromosomes,mu):
    mean = 0
    for i in chromosomes:
        mean += i[-1]
    return mean/mu
    

In [30]:
getMeanT(initial2,100)

0.035069983371305645

In [31]:
def parentsSelectionT(chromosomes,mu):
    r = random.random()
    mean = getMeanT(chromosomes,mu)
    cont = 0
    parents = []
    for i in range(mu):
        cont += (chromosomes[i][-1] / mean )
        while cont > r:
            parents.append(chromosomes[i])
            r+=1
    return parents

In [32]:
arr = parentsSelectionT(initial2,100)
arr

[[array([0.83071645, 0.06269388]), 5.764199474317527, 0.14783715409293055],
 [array([0.83071645, 0.06269388]), 5.764199474317527, 0.14783715409293055],
 [array([0.83071645, 0.06269388]), 5.764199474317527, 0.14783715409293055],
 [array([0.83071645, 0.06269388]), 5.764199474317527, 0.14783715409293055],
 [array([-2.65623267,  0.03454488]), 68.45023151559613, 0.014398800092918832],
 [array([3.752152  , 0.87874363]), 5.052004191276636, 0.1652345187469303],
 [array([3.752152  , 0.87874363]), 5.052004191276636, 0.1652345187469303],
 [array([3.752152  , 0.87874363]), 5.052004191276636, 0.1652345187469303],
 [array([3.752152  , 0.87874363]), 5.052004191276636, 0.1652345187469303],
 [array([-1.15735031,  1.2734984 ]), 5.680287107391992, 0.14969416492495688],
 [array([-1.15735031,  1.2734984 ]), 5.680287107391992, 0.14969416492495688],
 [array([-1.15735031,  1.2734984 ]), 5.680287107391992, 0.14969416492495688],
 [array([-1.15735031,  1.2734984 ]), 5.680287107391992, 0.14969416492495688],
 [arr

In [33]:
def combinationT(chro1,chro2,variables,bounds):
    chro1,chro2 = chro1[0],chro2[0]
    
    alpha = random.uniform(0,1)
    son1 = alpha * np.array(chro1) + (1-alpha) * np.array(chro2)
    alpha = random.uniform(0,1)
    son2 = alpha * np.array(chro2) + (1-alpha) * np.array(chro1)
    
    son1 = getValuesT([son1], variables, bounds)
    son2 = getValuesT([son2], variables, bounds)
    return son1[0],son2[0]


In [34]:
h1,h2 = combinationT(initial2[0],initial2[1],2,[-4.5,4.5])
print(h1)
print()
print(h2)

[array([ 2.50537226, -0.43799004]), 4.480343671500959, 0.18247030842248613]

[array([ 2.3633718 , -0.39553514]), 3.312449561356102, 0.23188677009952968]


In [35]:
def mutationT(arr,pm):
    chrom = arr[0]
    newchrom = []
    for i in chrom:
        if np.random.uniform(0,1,1)[0] < pm:
            newchrom.append(np.random.uniform(-4.5,4.5))
        else:
            newchrom.append(i)
    return getValuesT([newchrom],2, [-4.5, 4.5])

In [36]:
mutationT([[-3.13434993, -1.0947048 ],165.14480204126107,0.006018846137308923],0.90)

[[[2.2770426894778844, 4.271188815934183],
  33407.03366360735,
  2.9932920029631626e-05]]

In [37]:
def beestworstT(arr):
    best = arr[0][-1]
    best_complete_array = arr[0]
    worst =arr[0][-1]
    worst_complete_array = arr[0]
    for i in arr:
        if i[-1] > best:
            best = i[-1]
            best_complete_array = i
            
        if i[-1] < worst:
            worst = i[-1]
            worst_complete_array = i
    return best_complete_array,worst_complete_array

In [38]:
def survivorseriesT(parents, son):
    bestB, worstB =  beestworstT(parents)
    bestW, worstW =  beestworstT(son)
    if bestB[-1] > worstW[-1]:
        son.remove(worstW)
        son.append(bestB)
    return son

In [39]:
def AGT(G,mu,pc,pm,bounds,variables):
    initial = initialPopuT(mu,variables,bounds)
    initial = getValuesT(initial, variables, bounds)
    t = 0
    while t<=G:
        parents = parentsSelectionT(initial,mu)
        sons = []
        
        for i in range(0,mu,2):
            if random.random() < pc:
                h1,h2= combinationT(parents[i],parents[i+1],variables,bounds)
            else:
                h1 = parents[i].copy()
                h2 = parents[i+1].copy()
                
               
           # if len(h1[0][0]) == 1:
            #    h1 = [h1]
                
            #if len(h2[0][0]) == 1:
             #   h2 = [h2]
                
            h1 = mutationT(h1,pm)
            h2 = mutationT(h2,pm)
            h1 = h1[0]
            h2 = h2[0]

            sons.append(h1)
            sons.append(h2)
        
        initial = survivorseriesT(parents, sons)
        t +=1
        
    best,_ = beestworstT(initial)  
    return best
    

In [45]:
AGT(100,100,0.9,0.0001,[-4.5, 4.5],2)

[[2.83754021183263, 0.46378551257528144],
 0.005958094869357692,
 0.9940771937720414]

In [51]:
M = int(input()) 

def std(solutions,mean):
    m = mean
    all_ = 0
    for i in range(len(solutions)):
        all_ = all_ + (solutions[i][1] - m)**2
    return np.sqrt(all_/len(solutions))



Mean = 0
curr = []
for i in range(M):
    a = AGT(100,100,0.9,0.0001,[-4.5, 4.5],2)
    Mean += a[1]
    curr.append(a)
Mean = Mean/M    
curr = sorted(curr, key = lambda x:x[1])
print("Best Solution Xs: {}, F: {}".format(curr[0][0],curr[0][1]))  
print("Worst Solution Xs: {}, F: {}".format(curr[-1][0],curr[-1][1]))  
print("Median Solution Xs: {}, F: {}".format(curr[M//2][0],curr[M//2][1]))  
print("Function Mean: {}".format(Mean))
print("Standard Deviation: {}".format(std(curr,Mean)))

 100


Best Solution Xs: [3.00084108023354, 0.5002080872321268], F: 1.1312772359784502e-07
Worst Solution Xs: [1.8277007892660264, 0.08017174700436674], F: 0.8584082169332921
Median Solution Xs: [2.8124199712380307, 0.40918782895271455], F: 0.034522665305681245
Function Mean: 0.11592103827120072
Standard Deviation: 0.20353305097668029


In [52]:
M = int(input()) 

def std(solutions,mean):
    m = mean
    all_ = 0
    for i in range(len(solutions)):
        all_ = all_ + (solutions[i][-2] - m)**2
    return np.sqrt(all_/len(solutions))



Mean = 0
curr = []
for i in range(M):
    a = AG(100,100,0.9,0.0001,[-4.5, 4.5],2)
    Mean += a[-2]
    curr.append(a)
Mean = Mean/M    
curr = sorted(curr, key = lambda x:x[-2])
print("Best Solution Xs: {}, F: {}".format(curr[0][2],curr[0][-2]))  
print("Worst Solution Xs: {}, F: {}".format(curr[-1][2],curr[-1][-2]))  
print("Median Solution Xs: {}, F: {}".format(curr[M//2][2],curr[M//2][-2]))  
print("Function Mean: {}".format(Mean))
print("Standard Deviation: {}".format(std(curr,Mean)))

 100


Best Solution Xs: [3.0013733748397726, 0.5012818165171211], F: 2.0813055099791448e-05
Worst Solution Xs: [1.6675517304522982, -0.0634499175975094], F: 1.3377276068451442
Median Solution Xs: [2.8546969419520236, 0.40789232741256143], F: 0.05433650147224939
Function Mean: 0.17847111433557003
Standard Deviation: 0.2794965281220474


## 10 Points
## 10 Puntos

### Ackley

In [40]:
def f(arrX):
    firstpart = 0.0
    secondpart = 0.0
    for c in arrX:
        firstpart += c**2.0
        secondpart += math.cos(2.0*math.pi*c)
    n = float(len(arrX))
    return (-20.0*math.exp(-0.2*math.sqrt(firstpart/n)) - math.exp(secondpart/n) + 20 + math.e)

In [41]:
def RealF(x1):
    
    value = f(x1)
    return 1/(value+1)
    

In [42]:
def getValuesT(chromosomes, variables, bounds):
    c=0
    valuesR = []
    valuesN = []
    for i in range(len(chromosomes)):

#        newx = bounds[0] + ((9*chromosomes[0])/((2**num_bits)-1))
#        valuesR.append(newx)
            
#            curr = ""
#            c = 0

     chromosomes[i]= [chromosomes[i],f(chromosomes[i]),RealF(chromosomes[i])]
#    valuesR = []
#    valuesN = []
    
    return chromosomes

In [43]:
initial = initialPopuT(100,2,[-4.5,4.5])

initial2 = getValuesT(initial, 2, [-4.5,4.5])
initial2

[[array([ 0.55688453, -3.19325234]), 9.327148806462665, 0.09683214784067082],
 [array([-3.48219011, -3.15659944]), 11.626969365158425, 0.07919556712945691],
 [array([-0.95880995,  3.25928394]), 8.773435581543039, 0.10231816556794854],
 [array([-2.29261873,  2.76742415]), 9.761785050442697, 0.09292138760556864],
 [array([1.48252306e+00, 2.72014747e-04]),
  5.498075502474318,
  0.15389171757379902],
 [array([-2.83188386, -0.01506982]), 7.21458040471666, 0.12173476315671809],
 [array([3.20286229, 4.15030905]), 11.638099009603877, 0.07912582416390988],
 [array([ 0.06665825, -3.33379812]), 9.009613754469866, 0.09990395479080726],
 [array([-3.63884975,  0.15106958]), 9.798784845641745, 0.09260301175493718],
 [array([-1.07284092, -2.95517798]), 7.3652623416109115, 0.11954197718650728],
 [array([3.34172005, 4.19154102]), 12.435928088563658, 0.0744273111175086],
 [array([ 1.35790596, -0.47418642]), 5.9532222725929245, 0.14381821273593348],
 [array([-3.62115805,  4.01687259]), 12.265221490784056

In [44]:
def getValues(chromosomes,num_bits, variables, bounds):
    c=0
    curr = ""
    valuesR = []
    valuesN = []
    for chromosome in range(len(chromosomes)):
        for i in chromosomes[chromosome]:
            curr = curr+i
            c+=1
            if c == num_bits:
                x = int(curr,base=2)
                valuesN.append(x)
                newx = bounds[0] + ((9*x)/((2**num_bits)-1))
                valuesR.append(newx)
                curr = ""
                c = 0
        chromosomes[chromosome] = [chromosomes[chromosome],valuesN,valuesR,f(valuesR),RealF(valuesR)]
        valuesR = []
        valuesN = []
    
    return chromosomes

In [45]:
def AGT(G,mu,pc,pm,bounds,variables):
    initial = initialPopuT(mu,variables,bounds)
    initial = getValuesT(initial, variables, bounds)
    t = 0
    while t<=G:
        parents = parentsSelectionT(initial,mu)
        sons = []
        
        for i in range(0,mu,2):
            if random.random() < pc:
                h1,h2= combinationT(parents[i],parents[i+1],variables,bounds)
            else:
                h1 = parents[i].copy()
                h2 = parents[i+1].copy()
                
               
           # if len(h1[0][0]) == 1:
            #    h1 = [h1]
                
            #if len(h2[0][0]) == 1:
             #   h2 = [h2]
                
            h1 = mutationT(h1,pm)
            h2 = mutationT(h2,pm)
            h1 = h1[0]
            h2 = h2[0]

            sons.append(h1)
            sons.append(h2)
        
        initial = survivorseriesT(parents, sons)
        t +=1
        
    best,_ = beestworstT(initial)  
    return best
    

In [50]:
AGT(100,100,0.9,0.0001,[-30, 30],2)

[[-0.0012117766435778784, 0.0009101141133350586],
 0.004347610117758638,
 0.9956712097744237]

In [61]:
AG(100,100,0.9,0.0001,[-30, 30],2)

['10000000000000100000000000011001',
 [8192, 8193],
 [0.0002746749679545246, 0.0008240249038635739],
 0.002476857617122352,
 0.9975292620489915]

# 20 Points
# 20 Puntos

### Ackley
### Real

M = 100 <br>
Variables = 5

In [56]:
M = int(input()) 

def std(solutions,mean):
    m = mean
    all_ = 0
    for i in range(len(solutions)):
        all_ = all_ + (solutions[i][1] - m)**2
    return np.sqrt(all_/len(solutions))



Mean = 0
curr = []
for i in range(M):
    a = AGT(100,200,1.22,0.0095,[-30, 30],5)
    Mean += a[1]
    curr.append(a)
Mean = Mean/M    
curr = sorted(curr, key = lambda x:x[1])
print("Best Solution Xs: {}, F: {}".format(curr[0][0],curr[0][1]))  
print("Worst Solution Xs: {}, F: {}".format(curr[-1][0],curr[-1][1]))  
print("Median Solution Xs: {}, F: {}".format(curr[M//2][0],curr[M//2][1]))  
print("Function Mean: {}".format(Mean))
print("Standard Deviation: {}".format(std(curr,Mean)))

 100


Best Solution Xs: [3.2132441011806986e-06, 0.0038927105495925405, -7.59497010194072e-05, -0.0017214041544693194, -0.0002207071006456714], F: 0.007818943446960436
Worst Solution Xs: [-0.015849606719664358, -0.04149162241083374, 0.0020617638091155743, 0.06544239617822636, 0.012012421666916441], F: 0.20977189309836008
Median Solution Xs: [-0.004162982963660862, -0.004882108610330452, -0.01648755386765317, -0.014530326979374238, -0.00865645472175322], F: 0.05015274807283321
Function Mean: 0.06443123392410109
Standard Deviation: 0.04450143154986199


M = 100 <br>
Variables = 10

In [76]:
M = int(input()) 

def std(solutions,mean):
    m = mean
    all_ = 0
    for i in range(len(solutions)):
        all_ = all_ + (solutions[i][1] - m)**2
    return np.sqrt(all_/len(solutions))



Mean = 0
curr = []
for i in range(M):
    a = AGT(100,600,1.22,0.0200,[-30, 30],10)
    Mean += a[1]
    curr.append(a)
Mean = Mean/M    
curr = sorted(curr, key = lambda x:x[1])
print("Best Solution Xs: {}, F: {}".format(curr[0][0],curr[0][1]))  
print("Worst Solution Xs: {}, F: {}".format(curr[-1][0],curr[-1][1]))  
print("Median Solution Xs: {}, F: {}".format(curr[M//2][0],curr[M//2][1]))  
print("Function Mean: {}".format(Mean))
print("Standard Deviation: {}".format(std(curr,Mean)))

 100


Best Solution Xs: [0.0026690955107140334, -0.0037115388985214104, 0.024277551255858634, -0.011648806157072096, 0.023347646658296296, -0.011127845160303265, 0.006150141824306206, -0.002755018020146589, -0.015356967205936777, 0.005200565358861012], F: 0.06165755004353857
Worst Solution Xs: [0.08785158825692317, 0.003890704420749626, -0.11811174169099267, 0.034567300804562655, -0.07272216976881744, 0.023508355871014554, -0.009793268179928524, 0.1063307378068605, -0.1344978150580062, 0.012612579105641411], F: 0.5876725470588195
Median Solution Xs: [-0.021865396487567357, 0.058092665980192464, -0.01789015755216901, 0.04741621460709282, -0.03257154610090206, -0.04090675336643212, 0.00525422860592599, 0.020660718872562996, 0.05717107007166153, 0.03588685035568058], F: 0.2243783451004444
Function Mean: 0.23318064992699156
Standard Deviation: 0.09526713615892675


M = 100 <br>
Variables = 20

In [77]:
M = int(input()) 

def std(solutions,mean):
    m = mean
    all_ = 0
    for i in range(len(solutions)):
        all_ = all_ + (solutions[i][1] - m)**2
    return np.sqrt(all_/len(solutions))



Mean = 0
curr = []
for i in range(M):
    a = AGT(100,600,1.22,0.0200,[-30, 30],20)
    Mean += a[1]
    curr.append(a)
Mean = Mean/M    
curr = sorted(curr, key = lambda x:x[1])
print("Best Solution Xs: {}, F: {}".format(curr[0][0],curr[0][1]))  
print("Worst Solution Xs: {}, F: {}".format(curr[-1][0],curr[-1][1]))  
print("Median Solution Xs: {}, F: {}".format(curr[M//2][0],curr[M//2][1]))  
print("Function Mean: {}".format(Mean))
print("Standard Deviation: {}".format(std(curr,Mean)))

 100


Best Solution Xs: [0.14346950922897825, 0.16493242438091135, -0.014625418462798176, 0.05837487973783641, -0.21342711541537118, -0.018035109725118806, 0.0925023008981545, 0.0208311399546476, -0.006842567022348575, 0.24026460890996398, 0.45206049827690475, -0.09072295735896671, -0.10711431822929746, -0.1372506286295645, 0.07433072201929272, 0.19375675797356356, -0.13347520438356325, -0.0363838845197482, -0.14759984319552782, 0.004189541571234618], F: 1.422752639587753
Worst Solution Xs: [-0.3593339410181551, -0.05028811602184269, 0.3417529825916735, -0.038317092961792504, -0.9478636794258062, -0.0711592800094733, 1.0246184848410476, -0.0041060737580359, 0.8866282035563975, 0.04893250045857234, -0.19902561105667407, -0.4379899588935554, 0.3073663899563549, 0.28553220374673893, -0.06576632738831673, -0.00553418106142288, 0.09340126296101786, 0.020293746848068527, 0.12527246037291734, 0.8364358866888593], F: 2.8011952847055785
Median Solution Xs: [-0.13571878671684107, 0.24254033437732597, 

### Ackley
### Chromosomes

M = 100 <br>
Variables = 5

In [78]:
M = int(input()) 

def std(solutions,mean):
    m = mean
    all_ = 0
    for i in range(len(solutions)):
        all_ = all_ + (solutions[i][-2] - m)**2
    return np.sqrt(all_/len(solutions))



Mean = 0
curr = []
for i in range(M):
    a = AG(100,200,1.22,0.0095,[-30, 30],5)
    Mean += a[-2]
    curr.append(a)
Mean = Mean/M    
curr = sorted(curr, key = lambda x:x[-2])
print("Best Solution Xs: {}, F: {}".format(curr[0][2],curr[0][-2]))  
print("Worst Solution Xs: {}, F: {}".format(curr[-1][2],curr[-1][-2]))  
print("Median Solution Xs: {}, F: {}".format(curr[M//2][2],curr[M//2][-2]))  
print("Function Mean: {}".format(Mean))
print("Standard Deviation: {}".format(std(curr,Mean)))

 100


Best Solution Xs: [-0.0002746749679545246, 0.0002746749679545246, 0.0002746749679545246, -0.0002746749679545246, -0.0002746749679545246], F: 0.0011027178938012305
Worst Solution Xs: [-0.0019227247756816723, 0.01785387291704854, 0.010162973814320075, 0.0041201245193187575, 0.004669474455227807], F: 0.043494087078780996
Median Solution Xs: [-0.001373374839772623, -0.0024720747115916097, 0.001373374839772623, -0.005218824391136856, -0.0002746749679545246], F: 0.011305939436497159
Function Mean: 0.012692121015311794
Standard Deviation: 0.008153768887046504


M = 100 <br>
Variables = 10

In [79]:
M = int(input()) 

def std(solutions,mean):
    m = mean
    all_ = 0
    for i in range(len(solutions)):
        all_ = all_ + (solutions[i][-2] - m)**2
    return np.sqrt(all_/len(solutions))



Mean = 0
curr = []
for i in range(M):
    a = AG(100,600,1.22,0.0200,[-30, 30],10)
    Mean += a[-2]
    curr.append(a)
Mean = Mean/M    
curr = sorted(curr, key = lambda x:x[-2])
print("Best Solution Xs: {}, F: {}".format(curr[0][2],curr[0][-2]))  
print("Worst Solution Xs: {}, F: {}".format(curr[-1][2],curr[-1][-2]))  
print("Median Solution Xs: {}, F: {}".format(curr[M//2][2],curr[M//2][-2]))  
print("Function Mean: {}".format(Mean))
print("Standard Deviation: {}".format(std(curr,Mean)))

 100


Best Solution Xs: [0.00851492400659204, -0.009613623878410138, -0.009064273942501089, 0.01840322285295759, 0.003570774583409708, -0.01785387291704854, 0.011811023622047223, -0.017304522981138604, 0.027192821827504154, 0.009064273942501089, 0.003021424647500659], F: 0.0669537085966323
Worst Solution Xs: [-0.04642006958432532, 0.0749862662516021, -0.9451565647317342, -0.9891045596044679, -0.9358176158212781, 0.03378502105841452, 0.006866874198864892, 1.007233107489471, 0.0024720747115916097, 1.0390954037722029, -1.0550265519135689], F: 2.8372094350560393
Median Solution Xs: [-0.03543307086614167, -0.014557773301593357, 0.14695110785570442, 0.07114081670023786, 0.00796557407068299, 0.06509796740523743, 0.027192821827504154, -0.0865226149056948, -0.026643471891595105, -0.054110968687053784, 0.09641091375206035], F: 0.5169609054017283
Function Mean: 0.787605261139611
Standard Deviation: 0.6816173132540526


M = 100 <br>
Variables = 20

In [74]:
M = int(input()) 

def std(solutions,mean):
    m = mean
    all_ = 0
    for i in range(len(solutions)):
        all_ = all_ + (solutions[i][-2] - m)**2
    return np.sqrt(all_/len(solutions))



Mean = 0
curr = []
for i in range(M):
    a =  AG(100,400,2.5,0.045,[-30, 30],20)
    Mean += a[-2]
    curr.append(a)
Mean = Mean/M    
curr = sorted(curr, key = lambda x:x[-2])
print("Best Solution Xs: {}, F: {}".format(curr[0][2],curr[0][-2]))  
print("Worst Solution Xs: {}, F: {}".format(curr[-1][2],curr[-1][-2]))  
print("Median Solution Xs: {}, F: {}".format(curr[M//2][2],curr[M//2][-2]))  
print("Function Mean: {}".format(Mean))
print("Standard Deviation: {}".format(std(curr,Mean)))

 100


Best Solution Xs: [-1.2049990844167735, -0.9797656106940122, -0.09860831349569654, 0.35954953305255444, 0.8286943783189891, 1.0330525544772016, -0.02499542208386707, 0.5869804065189523, 0.12387841054751902, 0.20298480131843988, -0.11344076176524442, -0.03433437099432357, 0.319996337667094, 0.975370811206739, 0.1963926020875295, -0.10629921259842501, -1.1885185863394985, -0.0519135689434167, -1.3060794726240617, -0.8946163706280901, 0.9143929683208203, 0.1315693096502475], F: 3.6801808055929155
Worst Solution Xs: [1.4274858084599886, 0.483702618568028, -0.22331074894707914, -0.7479399377403406, 0.3304339864493686, -1.6812854788500275, -0.34361838491118846, -0.048617469327961516, 2.2624977110419335, -0.8979124702435453, -0.17277055484343506, 1.8691631569309655, -1.3005859732649698, -0.7787035341512545, -1.729628273210035, 1.4774766526277237, 1.0385460538362938, -2.0037538912287127, -1.5368064457059147, -0.0980589635597875, 1.6779893792345728, -0.7600256363303424], F: 6.113106708109356
Me

## Ackley Function
## Función de Ackley

### AG Chromosomes
### M = 100

| Variables | Best Solution  | Worst Solution    | Median Solution | Function Mean | Standard Deviation |
| :---      |    :----:      |          ---:     |                ---: |          ---: |               ---: |
| 5         | 0.45840928833040495  | 3.550733934318043 | 2.0941315378342122 | 2.03687809929765 |0.7426478356804945|
| 10        | 1.9621620271137918   |5.576046831285826  | 3.5545452813970937 | 3.5993247563793505|0.5904238732082906|
| 20        |   4.010357781885304  | 6.287537430599009   |  5.111737947287647 | 5.129792321636037|0.46915084737940854 |

## Ackley Function
## Función de Ackley

### AG Real
### M = 100

| Variables | Best Solution  | Worst Solution    | Median Solution | Function Mean | Standard Deviation |
| :---      |    :----:      |          ---:     |                ---: |          ---: |               ---: |
| 5         | 2.0702519291089563  | 8.847238533669787 |  4.111777994982052 | 4.27133178513078 |1.2157724296526173|
| 10        | 3.653367954401564   |9.556601948159734  | 6.461780721662116 | 6.3536881561542215|1.190910804965387|
| 20        |   5.353412890450137 | 10.014191595794967   |   7.698642424972599| 7.646131180358376|0.8919945822166411 |

## Beale Function
## Función de Beale

### M = 100
### Variables = 2


| Variables | Best Solution  | Worst Solution    | Median Solution | Function Mean | Standard Deviation |
| :---      |    :----:      |          ---:     |                ---: |          ---: |               ---: |
|AG Real| 1.1312772359784502e-07  | 0.8584082169332921|  0.034522665305681245| 0.11592103827120072 |0.20353305097668029|
| AG Chromosomes        | 2.0813055099791448e-05  |1.3377276068451442| 0.05433650147224939 |  0.17847111433557003|0.2794965281220474|


In [ ]:
#######################################################################################################################################################################################################################################################################

## 40 Points
## 40 Puntos

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import random
import math

In [2]:
def custominput():
    print("Txt name")
    txt = "Texto_input.txt"#input()
    #print("M iterations")
    #M = int(input())
        
    with open(txt, 'r') as file:
         input_lines = [line.strip() for line in file]
    global cities 
    cities = int(input_lines[0])
    #print(input_lines[1].split())
    pc,pm, maxiter = input_lines[1].split()
    pc,pm, maxiter = float(pc),float(pm), float(maxiter)




    cost_city = [] 
    for i in range(2,cities+1):
        costos_ni = list(map(int,input_lines[i].split()))
        cost_city.append(costos_ni)

    new_cost = np.zeros((cities, cities))
    for i in range(len(cost_city)+1):
        new_cost[i][i] = float('inf')
    c = 1
    for i in range(len(cost_city)):
        new_cost[i][c:] = cost_city[i]
        c+=1
    dic = {idx + 1 : new_cost[idx] for idx in range(len(new_cost))} 
    #######################################################
    for j in range(cities):
        count=0
        for i in range(len(new_cost[j])):#sera j
            if new_cost[i][j] == 0:
                new_cost[i][j] =new_cost[j][count]
            count+=1
            
    return new_cost

In [3]:
dic = custominput()
dic

Txt name


array([[inf, 49., 30., 53., 72., 19., 76., 87., 45., 48.],
       [49., inf, 19., 38., 32., 31., 75., 69., 61., 25.],
       [30., 19., inf, 41., 98., 56.,  6.,  6., 45., 53.],
       [53., 38., 41., inf, 52., 29., 46., 90., 23., 98.],
       [72., 32., 98., 52., inf, 63., 90., 69., 50., 82.],
       [19., 31., 56., 29., 63., inf, 60., 88., 41., 95.],
       [76., 75.,  6., 46., 90., 60., inf, 61., 92., 10.],
       [87., 69.,  6., 90., 69., 88., 61., inf, 82., 73.],
       [45., 61., 45., 23., 50., 41., 92., 82., inf,  5.],
       [48., 25., 53., 98., 82., 95., 10., 73.,  5., inf]])

In [4]:
def powa(new_dict,target):

    global cities
    for i in range(cities):
        new_dict[i][target]  = float('inf')

In [5]:
def gettotal(x,new_dict):
    back = new_dict
    total = 0
    currcity = 0
    co = 0
    for i in x:
        if co == 9:
            break
        total = total + new_dict[i][x[co+1]]
        co+=1
    lastcity = x[-2]
    destiny = x[-1]
    total += new_dict[lastcity][destiny]
    new_dict = back

    return total

In [6]:
gettotal([0, 5, 3, 8, 9, 6, 2, 7, 1, 4, 0],dic.copy())

271.0

In [115]:
def initialPopuTravel(popu,variables,dic):
    Pupulation =[]
    for i in range(popu):
        individual = [0]
        cities = list(range(1, variables))
        for i in range(variables):
            if len(cities) == 0:
                break
            target = random.choice(cities)
            individual.append(target)
            cities.remove(target)
        individual.append(0)##
        total = gettotal(individual,dic.copy())
        Pupulation.append([individual,total])
    return Pupulation
            

In [116]:
initialpopu= initialPopuTravel(100,10,dic.copy())
initialpopu

[[[0, 5, 2, 8, 6, 9, 3, 7, 4, 1, 0], 560.0],
 [[0, 3, 6, 2, 9, 1, 8, 4, 7, 5, 0], 470.0],
 [[0, 4, 8, 2, 1, 7, 3, 5, 9, 6, 0], 555.0],
 [[0, 1, 3, 5, 9, 2, 7, 4, 8, 6, 0], 557.0],
 [[0, 7, 8, 4, 9, 6, 2, 3, 1, 5, 0], 446.0],
 [[0, 7, 4, 1, 9, 2, 6, 3, 8, 5, 0], 401.0],
 [[0, 1, 5, 9, 8, 7, 3, 6, 4, 2, 0], 616.0],
 [[0, 5, 7, 9, 1, 4, 8, 3, 2, 6, 0], 433.0],
 [[0, 4, 3, 2, 9, 6, 1, 5, 7, 8, 0], 549.0],
 [[0, 5, 2, 6, 7, 4, 3, 8, 9, 1, 0], 365.0],
 [[0, 9, 5, 6, 8, 7, 3, 2, 4, 1, 0], 687.0],
 [[0, 1, 5, 9, 8, 6, 2, 3, 7, 4, 0], 550.0],
 [[0, 6, 7, 9, 3, 5, 2, 1, 4, 8, 0], 539.0],
 [[0, 7, 5, 3, 1, 9, 8, 6, 2, 4, 0], 540.0],
 [[0, 8, 1, 9, 3, 7, 6, 5, 4, 2, 0], 631.0],
 [[0, 6, 5, 1, 9, 3, 7, 4, 2, 8, 0], 637.0],
 [[0, 4, 7, 5, 6, 2, 9, 1, 8, 3, 0], 510.0],
 [[0, 6, 8, 5, 9, 3, 7, 4, 1, 2, 0], 642.0],
 [[0, 6, 8, 4, 3, 2, 1, 5, 9, 7, 0], 616.0],
 [[0, 6, 1, 8, 5, 2, 4, 7, 3, 9, 0], 712.0],
 [[0, 1, 6, 4, 9, 3, 7, 5, 8, 2, 0], 688.0],
 [[0, 9, 2, 1, 6, 8, 4, 5, 7, 3, 0], 631.0],
 [[0, 4, 7

In [117]:
initialpopu[0][-1]

560.0

In [118]:
def getMeanTravel(chromosomes,mu):
    mean = 0
    for i in chromosomes:
        mean += i[-1]
    return mean/mu
    

In [119]:
getMeanTravel(initialpopu,100)

545.01

In [120]:
def parentsSelectionTravel(chromosomes,mu):
    r = random.random()
    #print("chr",chromosomes)
    mean = getMeanTravel(chromosomes,mu)
    cont = 0
    parents = []
    for i in range(mu):
        cont += (chromosomes[i][-1] / mean )
        while cont > r:
            parents.append(chromosomes[i])
            r+=1
    #print("parents",parents)
    return parents

In [123]:
parentsSelectionTravel([[[0, 5, 2, 8, 6, 9, 3, 7, 4, 1, 0], 560.0],
 [[0, 3, 6, 2, 9, 1, 8, 4, 7, 5, 0], 470.0],
 [[0, 4, 8, 2, 1, 7, 3, 5, 9, 6, 0], 555.0],
 [[0, 1, 3, 5, 9, 2, 7, 4, 8, 6, 0], 557.0],
 [[0, 7, 8, 4, 9, 6, 2, 3, 1, 5, 0], 446.0],
 [[0, 7, 4, 1, 9, 2, 6, 3, 8, 5, 0], 401.0],
 [[0, 1, 5, 9, 8, 7, 3, 6, 4, 2, 0], 616.0],
 [[0, 5, 7, 9, 1, 4, 8, 3, 2, 6, 0], 433.0],
 [[0, 4, 3, 2, 9, 6, 1, 5, 7, 8, 0], 549.0],
 [[0, 5, 2, 6, 7, 4, 3, 8, 9, 1, 0], 365.0],
 [[0, 9, 5, 6, 8, 7, 3, 2, 4, 1, 0], 687.0],
 [[0, 1, 5, 9, 8, 6, 2, 3, 7, 4, 0], 550.0],
 [[0, 6, 7, 9, 3, 5, 2, 1, 4, 8, 0], 539.0],
 [[0, 7, 5, 3, 1, 9, 8, 6, 2, 4, 0], 540.0],
 [[0, 8, 1, 9, 3, 7, 6, 5, 4, 2, 0], 631.0],
 [[0, 6, 5, 1, 9, 3, 7, 4, 2, 8, 0], 637.0],
 [[0, 4, 7, 5, 6, 2, 9, 1, 8, 3, 0], 510.0],
 [[0, 6, 8, 5, 9, 3, 7, 4, 1, 2, 0], 642.0],
 [[0, 6, 8, 4, 3, 2, 1, 5, 9, 7, 0], 616.0],
 [[0, 6, 1, 8, 5, 2, 4, 7, 3, 9, 0], 712.0],
 [[0, 1, 6, 4, 9, 3, 7, 5, 8, 2, 0], 688.0],
 [[0, 9, 2, 1, 6, 8, 4, 5, 7, 3, 0], 631.0],
 [[0, 4, 7, 5, 1, 9, 2, 3, 6, 8, 0], 562.0],
 [[0, 9, 3, 1, 6, 4, 2, 8, 5, 7, 0], 708.0],
 [[0, 4, 3, 8, 6, 2, 9, 7, 5, 1, 0], 539.0],
 [[0, 6, 2, 7, 3, 8, 4, 9, 5, 1, 0], 508.0],
 [[0, 2, 6, 7, 4, 8, 9, 1, 3, 5, 0], 332.0],
 [[0, 5, 8, 7, 1, 6, 3, 9, 2, 4, 0], 653.0],
 [[0, 9, 7, 3, 5, 4, 6, 8, 2, 1, 0], 598.0],
 [[0, 7, 6, 9, 2, 3, 5, 8, 1, 4, 0], 487.0],
 [[0, 8, 7, 2, 6, 1, 4, 5, 9, 3, 0], 555.0],
 [[0, 5, 2, 6, 9, 8, 4, 1, 3, 7, 0], 393.0],
 [[0, 2, 1, 5, 8, 7, 9, 4, 6, 3, 0], 547.0],
 [[0, 7, 4, 9, 3, 2, 8, 6, 1, 5, 0], 639.0],
 [[0, 1, 8, 7, 3, 6, 2, 5, 9, 4, 0], 639.0],
 [[0, 8, 4, 9, 2, 5, 6, 1, 7, 3, 0], 633.0],
 [[0, 5, 4, 3, 6, 1, 7, 2, 9, 8, 0], 433.0],
 [[0, 2, 3, 7, 4, 6, 8, 9, 1, 5, 0], 492.0],
 [[0, 2, 7, 9, 4, 8, 3, 1, 6, 5, 0], 456.0],
 [[0, 4, 2, 9, 1, 6, 8, 5, 7, 3, 0], 687.0],
 [[0, 1, 6, 5, 8, 3, 4, 2, 7, 9, 0], 525.0],
 [[0, 4, 9, 6, 3, 7, 8, 2, 5, 1, 0], 563.0],
 [[0, 6, 5, 9, 4, 8, 1, 7, 3, 2, 0], 654.0],
 [[0, 1, 5, 7, 6, 2, 4, 3, 8, 9, 0], 461.0],
 [[0, 8, 9, 7, 3, 4, 1, 5, 2, 6, 0], 466.0],
 [[0, 1, 3, 4, 6, 8, 9, 2, 5, 7, 0], 610.0],
 [[0, 7, 1, 9, 8, 5, 6, 2, 4, 3, 0], 496.0],
 [[0, 9, 2, 3, 6, 1, 4, 8, 7, 5, 0], 534.0],
 [[0, 2, 6, 7, 8, 4, 5, 3, 9, 1, 0], 493.0],
 [[0, 7, 9, 2, 1, 4, 8, 5, 6, 3, 0], 514.0],
 [[0, 7, 4, 5, 2, 8, 1, 3, 6, 9, 0], 523.0],
 [[0, 3, 4, 1, 7, 6, 9, 2, 8, 5, 0], 435.0],
 [[0, 6, 3, 5, 2, 9, 1, 4, 7, 8, 0], 513.0],
 [[0, 3, 7, 2, 5, 9, 1, 4, 6, 8, 0], 584.0],
 [[0, 9, 6, 1, 2, 3, 4, 5, 7, 8, 0], 523.0],
 [[0, 9, 7, 1, 3, 6, 8, 4, 2, 5, 0], 589.0],
 [[0, 1, 3, 8, 2, 9, 7, 4, 6, 5, 0], 519.0],
 [[0, 7, 6, 3, 8, 5, 1, 9, 4, 2, 0], 524.0],
 [[0, 8, 3, 6, 5, 2, 1, 7, 9, 4, 0], 545.0],
 [[0, 9, 6, 4, 3, 8, 2, 5, 7, 1, 0], 530.0],
 [[0, 2, 7, 9, 8, 4, 6, 1, 3, 5, 0], 415.0],
 [[0, 8, 6, 2, 4, 5, 9, 1, 7, 3, 0], 636.0],
 [[0, 1, 7, 2, 6, 9, 5, 8, 4, 3, 0], 431.0],
 [[0, 3, 1, 4, 9, 6, 7, 2, 5, 8, 0], 424.0],
 [[0, 7, 3, 8, 5, 1, 9, 2, 4, 6, 0], 614.0],
 [[0, 4, 7, 2, 5, 6, 9, 1, 8, 3, 0], 435.0],
 [[0, 4, 1, 6, 9, 3, 5, 2, 8, 7, 0], 586.0],
 [[0, 1, 2, 4, 7, 9, 6, 3, 5, 8, 0], 479.0],
 [[0, 1, 8, 2, 7, 6, 5, 9, 4, 3, 0], 564.0],
 [[0, 4, 3, 2, 9, 7, 1, 8, 6, 5, 0], 592.0],
 [[0, 5, 1, 7, 8, 4, 9, 3, 2, 6, 0], 554.0],
 [[0, 4, 8, 7, 6, 3, 9, 1, 5, 2, 0], 551.0],
 [[0, 3, 6, 8, 2, 1, 9, 4, 7, 5, 0], 538.0],
 [[0, 2, 4, 6, 1, 5, 3, 8, 7, 9, 0], 579.0],
 [[0, 3, 4, 9, 1, 2, 7, 6, 8, 5, 0], 450.0],
 [[0, 1, 5, 9, 6, 2, 3, 8, 7, 4, 0], 478.0],
 [[0, 7, 3, 4, 2, 6, 9, 8, 5, 1, 0], 469.0],
 [[0, 9, 1, 5, 3, 4, 8, 6, 7, 2, 0], 424.0],
 [[0, 9, 4, 6, 8, 1, 2, 7, 3, 5, 0], 536.0],
 [[0, 1, 4, 6, 5, 7, 8, 2, 9, 3, 0], 650.0],
 [[0, 7, 1, 9, 3, 2, 8, 6, 4, 5, 0], 629.0],
 [[0, 4, 3, 6, 1, 5, 9, 2, 7, 8, 0], 557.0],
 [[0, 4, 7, 2, 3, 8, 6, 1, 9, 5, 0], 517.0],
 [[0, 8, 6, 9, 3, 7, 5, 2, 4, 1, 0], 658.0],
 [[0, 5, 3, 9, 6, 8, 4, 2, 1, 7, 0], 571.0],
 [[0, 9, 2, 6, 3, 8, 4, 1, 7, 5, 0], 434.0],
 [[0, 6, 3, 1, 5, 4, 7, 8, 2, 9, 0], 551.0],
 [[0, 7, 3, 2, 9, 5, 4, 8, 1, 6, 0], 691.0],
 [[0, 6, 7, 3, 1, 2, 9, 8, 4, 5, 0], 474.0],
 [[0, 3, 7, 5, 8, 2, 1, 4, 6, 9, 0], 516.0],
 [[0, 5, 2, 4, 3, 9, 1, 7, 6, 8, 0], 615.0],
 [[0, 1, 7, 2, 5, 8, 9, 3, 4, 6, 0], 542.0],
 [[0, 8, 7, 9, 6, 4, 3, 1, 5, 2, 0], 507.0],
 [[0, 6, 4, 8, 2, 5, 7, 1, 3, 9, 0], 658.0],
 [[0, 8, 5, 4, 2, 3, 7, 1, 6, 9, 0], 580.0],
 [[0, 2, 3, 9, 5, 7, 6, 8, 1, 4, 0], 670.0],
 [[0, 5, 1, 7, 8, 3, 9, 4, 2, 6, 0], 584.0],
 [[0, 1, 9, 8, 2, 6, 3, 7, 5, 4, 0], 489.0],
 [[0, 5, 6, 3, 2, 4, 1, 9, 8, 7, 0], 495.0],
 [[0, 8, 4, 7, 3, 6, 1, 2, 9, 5, 0], 561.0]],100)

[[[0, 5, 2, 8, 6, 9, 3, 7, 4, 1, 0], 560.0],
 [[0, 3, 6, 2, 9, 1, 8, 4, 7, 5, 0], 470.0],
 [[0, 4, 8, 2, 1, 7, 3, 5, 9, 6, 0], 555.0],
 [[0, 1, 3, 5, 9, 2, 7, 4, 8, 6, 0], 557.0],
 [[0, 7, 8, 4, 9, 6, 2, 3, 1, 5, 0], 446.0],
 [[0, 7, 4, 1, 9, 2, 6, 3, 8, 5, 0], 401.0],
 [[0, 1, 5, 9, 8, 7, 3, 6, 4, 2, 0], 616.0],
 [[0, 5, 7, 9, 1, 4, 8, 3, 2, 6, 0], 433.0],
 [[0, 4, 3, 2, 9, 6, 1, 5, 7, 8, 0], 549.0],
 [[0, 9, 5, 6, 8, 7, 3, 2, 4, 1, 0], 687.0],
 [[0, 9, 5, 6, 8, 7, 3, 2, 4, 1, 0], 687.0],
 [[0, 1, 5, 9, 8, 6, 2, 3, 7, 4, 0], 550.0],
 [[0, 6, 7, 9, 3, 5, 2, 1, 4, 8, 0], 539.0],
 [[0, 7, 5, 3, 1, 9, 8, 6, 2, 4, 0], 540.0],
 [[0, 8, 1, 9, 3, 7, 6, 5, 4, 2, 0], 631.0],
 [[0, 6, 5, 1, 9, 3, 7, 4, 2, 8, 0], 637.0],
 [[0, 4, 7, 5, 6, 2, 9, 1, 8, 3, 0], 510.0],
 [[0, 6, 8, 5, 9, 3, 7, 4, 1, 2, 0], 642.0],
 [[0, 6, 8, 4, 3, 2, 1, 5, 9, 7, 0], 616.0],
 [[0, 6, 1, 8, 5, 2, 4, 7, 3, 9, 0], 712.0],
 [[0, 1, 6, 4, 9, 3, 7, 5, 8, 2, 0], 688.0],
 [[0, 1, 6, 4, 9, 3, 7, 5, 8, 2, 0], 688.0],
 [[0, 9, 2

In [124]:
arr = parentsSelectionTravel(initialpopu,100)
arr

[[[0, 5, 2, 8, 6, 9, 3, 7, 4, 1, 0], 560.0],
 [[0, 3, 6, 2, 9, 1, 8, 4, 7, 5, 0], 470.0],
 [[0, 4, 8, 2, 1, 7, 3, 5, 9, 6, 0], 555.0],
 [[0, 1, 3, 5, 9, 2, 7, 4, 8, 6, 0], 557.0],
 [[0, 7, 4, 1, 9, 2, 6, 3, 8, 5, 0], 401.0],
 [[0, 1, 5, 9, 8, 7, 3, 6, 4, 2, 0], 616.0],
 [[0, 5, 7, 9, 1, 4, 8, 3, 2, 6, 0], 433.0],
 [[0, 4, 3, 2, 9, 6, 1, 5, 7, 8, 0], 549.0],
 [[0, 5, 2, 6, 7, 4, 3, 8, 9, 1, 0], 365.0],
 [[0, 9, 5, 6, 8, 7, 3, 2, 4, 1, 0], 687.0],
 [[0, 1, 5, 9, 8, 6, 2, 3, 7, 4, 0], 550.0],
 [[0, 6, 7, 9, 3, 5, 2, 1, 4, 8, 0], 539.0],
 [[0, 7, 5, 3, 1, 9, 8, 6, 2, 4, 0], 540.0],
 [[0, 8, 1, 9, 3, 7, 6, 5, 4, 2, 0], 631.0],
 [[0, 6, 5, 1, 9, 3, 7, 4, 2, 8, 0], 637.0],
 [[0, 4, 7, 5, 6, 2, 9, 1, 8, 3, 0], 510.0],
 [[0, 6, 8, 5, 9, 3, 7, 4, 1, 2, 0], 642.0],
 [[0, 6, 8, 4, 3, 2, 1, 5, 9, 7, 0], 616.0],
 [[0, 6, 8, 4, 3, 2, 1, 5, 9, 7, 0], 616.0],
 [[0, 6, 1, 8, 5, 2, 4, 7, 3, 9, 0], 712.0],
 [[0, 1, 6, 4, 9, 3, 7, 5, 8, 2, 0], 688.0],
 [[0, 9, 2, 1, 6, 8, 4, 5, 7, 3, 0], 631.0],
 [[0, 4, 7

In [140]:
def combinationTravel(chro1,chro2,dic):
    chro1,chro2 = chro1[0],chro2[0]
    stop1 = random.randint(1,len(chro1)-1)
    stop2 = random.randint(1,len(chro1)-1)
    lenchr = len(chro1)-1
    
    if stop1 > stop2:
        stop2df = stop2
        stop2 = stop1
        stop1 = stop2df
    
    

    son1 = [0]*lenchr
    son2 = [0]*lenchr
    son1[stop1:stop2] = chro2[stop1:stop2]
    son2[stop1:stop2] = chro1[stop1:stop2]

    
    son1range = list(range(lenchr))
    son2range = list(range(lenchr))
    
    for i in son1:
        if i in son1range:
            son1range.remove(i)
            
    for i in son2:
        if i in son2range:
            son2range.remove(i)
            
            
            


    
    
    
    for i in range(1,lenchr):
        if son1[i] == 0:
            son1[i] = son1range[0]
            son1range.pop(0)
            
    for i in range(1,lenchr):
        if son2[i] == 0:
            son2[i] = son2range[0]
            son2range.pop(0)
            

    son1.append(0)
    son2.append(0)
#    res = []
#    for i in son1:
#        if i not in res:
#            res.append(i)
#            
#            
#    res2 = []
#    for i in son2:
#        if i not in res2:
#            res2.append(i)

            

    son1 = [son1,gettotal(son1,dic.copy())]
    son2 = [son2,gettotal(son2,dic.copy())]

    return son1,son2


In [142]:
h1,h2 = combinationTravel(arr[0],arr[1],dic.copy())
print(h1)
print()
print(h2)

[[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0], 480.0]

[[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0], 480.0]


In [144]:
def MutationTravel(son1, variables,dic):
    son1 = son1[0]
    times = random.randint(1,variables-1)
    for i in range(times):
        one = random.randint(1,variables-1)
        onevalue = son1[one]
        two = random.randint(1,variables-1)
        twovalue = son1[two]
        son1[one] = twovalue
        son1[two] = onevalue
        
    son1 = [son1,gettotal(son1,dic.copy())]
    return son1
        
    
        

In [145]:
MutationTravel(h1, 10,dic.copy())

[[0, 9, 6, 3, 2, 5, 7, 1, 8, 4, 0], 541.0]

In [146]:
def beestworstTravel(arr):
    best = arr[0][-1]
    best_complete_array = arr[0]
    worst =arr[0][-1]
    worst_complete_array = arr[0]
    for i in arr:
        if i[-1] > best:
            best = i[-1]
            best_complete_array = i
            
        if i[-1] < worst:
            worst = i[-1]
            worst_complete_array = i
    return best_complete_array,worst_complete_array

In [149]:
def AGT(G,mu,pc,pm,variables,dic):
    initial = initialPopuTravel(mu,variables,dic.copy())
    t = 0
    while t<=G:
        parents = parentsSelectionTravel(initial,mu)
        #print("###")
        #print(parents)
        sons = []
        #print(len(parents))
        for i in range(0,mu,2):
            if random.random() < pc:
                h1,h2= combinationTravel(parents[i],parents[i+1],dic.copy())
            else:
                h1 = parents[i].copy()
                h2 = parents[i+1].copy()
                
               
           # if len(h1[0][0]) == 1:
            #    h1 = [h1]
                
            #if len(h2[0][0]) == 1:
             #   h2 = [h2]
                
            h1 = MutationTravel(h1, variables,dic.copy())
            h2 = MutationTravel(h2, variables,dic.copy())
            #h1 = h1[0]
            #h2 = h2[0]
            
            

            sons.append(h1)
            sons.append(h2)
            #print(1,len(parents))
        #initial = survivorseriesT(parents, sons)
        parents = sorted(parents, key = lambda x:x[-1])
        sons = sorted(sons, key = lambda x:x[-1])
        
        
        if sons[-1][-1] < parents[0][-1]:
            initial = parents
        else:
            initial = sons

            
        #print("##############################################################################################################")
        #print(sons)
       # print("len",len(sons))
        #3print("******************************************************")
        #print(2,len(parents))
        #if parents[0][-1] < sons[-1][-1]:
        #    parents[0]= sons[-1]
        #    parents.sort(key=lambda x:x[-1])

        
        ##initial = parents
        ##parents =initial
        t +=1
        #print(21233,t)
        #print(len(parents))
        #print(initial)
        
        
    initial.sort(key=lambda x:x[1])
    #print(initial)
    return initial[0]
        
        
        
    #best,_ = beestworstT(initial)  
    #return best
    

In [150]:
def custominput():
    print("Txt name")
    txt ="Texto_input.txt" #input()
    print("M iterations")
    M =1 #int(input())
    print("G")
    G = 100#int(input())
        
    with open(txt, 'r') as file:
         input_lines = [line.strip() for line in file]
    global cities 
    cities = int(input_lines[0])
    #print(input_lines[1].split())
    pc,pm, maxiter = input_lines[1].split()
    pc,pm, maxiter = float(pc),float(pm), int(maxiter)




    cost_city = [] 
    for i in range(2,cities+1):
        costos_ni = list(map(int,input_lines[i].split()))
        cost_city.append(costos_ni)

    new_cost = np.zeros((cities, cities))
    for i in range(len(cost_city)+1):
        new_cost[i][i] = float('inf')
    c = 1
    for i in range(len(cost_city)):
        new_cost[i][c:] = cost_city[i]
        c+=1
    dic = {idx + 1 : new_cost[idx] for idx in range(len(new_cost))} 
    #######################################################
    for j in range(cities):
        count=0
        for i in range(len(new_cost[j])):#sera j
            if new_cost[i][j] == 0:
                new_cost[i][j] =new_cost[j][count]
            count+=1
            
    
    for i in range(M):
        print(AGT(G,maxiter,pc,pm,cities,new_cost.copy()))
        
        #AGT(G,mu,pc,pm,variables,dic)
        
    
    

In [221]:
custominput()

Txt name
M iterations
G
[[0, 2, 7, 4, 1, 5, 6, 9, 8, 3, 0], 319.0]


## 10 Points
## 10 Puntos

In [226]:
def custominput():
    print("Txt name")
    txt = input()
    print("M iterations")
    M = int(input())
    print("G")
    G = int(input())
        
    with open(txt, 'r') as file:
         input_lines = [line.strip() for line in file]
    global cities 
    cities = int(input_lines[0])
    #print(input_lines[1].split())
    pc,pm, maxiter = input_lines[1].split()
    pc,pm, maxiter = float(pc),float(pm), int(maxiter)




    cost_city = [] 
    for i in range(2,cities+1):
        costos_ni = list(map(int,input_lines[i].split()))
        cost_city.append(costos_ni)

    new_cost = np.zeros((cities, cities))
    for i in range(len(cost_city)+1):
        new_cost[i][i] = float('inf')
    c = 1
    for i in range(len(cost_city)):
        new_cost[i][c:] = cost_city[i]
        c+=1
    dic = {idx + 1 : new_cost[idx] for idx in range(len(new_cost))} 
    #######################################################
    for j in range(cities):
        count=0
        for i in range(len(new_cost[j])):#sera j
            if new_cost[i][j] == 0:
                new_cost[i][j] =new_cost[j][count]
            count+=1
            
    
#    for i in range(M):
#        print(AGT(G,maxiter,pc,pm,cities,new_cost.copy()))
        
    
    def get_median(n):
        n = sorted(solutions, key=lambda tup: tup[1])
        l = len(n)
        if l%2 == 0:
            return n[(l//2)]
        else:
            return n[(l//2)-1]

    def mean(solutions):
        mean = 0
        for i in range(len(solutions)):
            mean = mean +solutions[i][1]
        return (mean/len(solutions))

    def std(solutions):
        m = mean(solutions)
        all_ = 0
        for i in range(len(solutions)):
            all_ = all_ + (solutions[i][1] - m)**2
        return np.sqrt(all_/len(solutions))
    
    solutions =[]
    best = float('inf')
    best_index = 0
    worst = 0
    worst_index = 0
    for i in range(M):
        curr = AGT(G,maxiter,pc,pm,cities,new_cost.copy())
        if curr[1] < best:
            best = curr[1]
            best_index = curr[0]
        if curr[1] > worst:
            worst = curr[1]
            worst_index = curr[0]

        solutions.append(curr)



    print("Best Solution:",best_index,best)
    print("Worst Solution:",worst_index,worst)
    print("Median:",get_median(solutions))
    print("Mean: ",mean(solutions))
    print("Standard deviation:",std(solutions))
    

In [234]:
custominput()
#Txt name
# Texto_input.txt
#M iterations
# 100
#G
# 100

Txt name


 Texto_input.txt


M iterations


 100


G


 100


Best Solution: [0, 5, 3, 4, 1, 2, 7, 6, 9, 8, 0] 278.0
Worst Solution: [0, 7, 4, 3, 6, 2, 5, 1, 9, 8, 0] 422.0
Median: [[0, 2, 7, 4, 6, 1, 9, 8, 3, 5, 0], 371.0]
Mean:  366.81
Standard deviation: 27.2061371752772
